### ISSUE TITLE
Single file IntelliSense consumes massive CPU indexing all system includes

### ISSUE THREAD

**Issue by MichaelGrupp** [microsoft/vscode-cpptools#1109]
Hi guys,

I experience a high CPU load, apparently caused by the C++ extension. One core is always used 100% even when no files are being edited. The problem doesn't occur with `code --disable-extensions`. My issue looks similar to this one: https://github.com/Microsoft/vscode/issues/33176

Here's my configuration:

- VSCode Version: Code 1.17.0 (be377c0faf7574a59f84940f593a6849f12e4de7, 2017-10-04T23:33:45.303Z)
- OS Version: Linux x64 4.4.0-96-generic
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
gc-excelviewer|Gra|1.1.15
code-settings-sync|Sha|2.8.3
githistory|don|0.2.3
python|don|0.7.0
gitlens|eam|5.4.1
LogFileHighlighter|emi|1.2.0
terminal|for|0.0.10
vscode-auto-open-markdown-preview|hnw|0.0.4
cpptools|ms-|0.13.1
platformio-ide|pla|0.7.3
shebang-snippets|rpi|0.0.3
bash-beautify|sha|0.1.1
rst-vscode|tht|2.0.0
cmake|twx|0.0.17
cmake-tools|vec|0.10.2
gitblame|wad|2.2.0
debug|web|0.21.2

---

Steps to Reproduce:

1. open vscode with C++ extension
2. do nothing, the CPU load is always high 

Reproduces without extensions: No

**Comment by MichaelGrupp** [comment]
![vscode_cpp_cpu](https://user-images.githubusercontent.com/8985495/31309324-c169a840-ab84-11e7-81dd-10116f2b7af1.png)

**Comment by Irvel** [comment]
I'm experiencing the exact same issue in macOS.

- VSCode Version 1.17.0 (1.17.0) (be377c0faf7574a59f84940f593a6849f12e4de7, 2017-10-05T06:03:48.991Z)

- macOS 10.13
- Microsoft C/C++ for VSCode Version 0.13.1

<img width="533" alt="screen shot 2017-10-08 at 10 04 13 am" src="https://user-images.githubusercontent.com/2577338/31318027-123b05fc-ac11-11e7-8dab-254bea657cc7.png">

**Comment by sean-mcmanus** [comment]
What are your c_cpp_properties.json settings? Depending on your browse settings, this might be "by design". You can enable logging to see if there's unexpected work occurring: https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/Enabling%20logging.md .

**Comment by bobbrow** [comment]
An hour and a half of parsing seems too much, even for large projects.  Can you give us more information about the project?  How many files?  Is it open source so we can try to debug the problem locally?  As Sean mentions, the "browse.path" setting may have an impact because we will recursively scan folders, so try not to include the same subfolders multiple times until we add support to prune rescans.

**Comment by MichaelGrupp** [comment]
ahh, interesting hint. When I experienced this, I had no folder selected (just opened a single file for quick editing). 

I just opened a random folder and the problem disappeared. It seems to re-appear when the folder is closed with `ctrl + kf`. As mentioned, this happens without opened files, too.

Here are some logs generated with the env. variables from the link of @sean-mcmanus posted above at loglevel 5.

* no folder selected on startup: [vscode.cpp.log.6113_no_folder_opened.txt](https://github.com/Microsoft/vscode-cpptools/files/1370060/vscode.cpp.log.6113_no_folder_opened.txt)
* startup with a folder selected: [vscode.cpp.log.6470_folder_preselected.txt](https://github.com/Microsoft/vscode-cpptools/files/1370059/vscode.cpp.log.6470_folder_preselected.txt)

Looking at the logs it seems like the recursion runs amok when there's no folder selected? :thinking:

**Comment by bobbrow** [comment]
Thank you for the logs.  That was super helpful.  It looks like VS Code is not sending us a root path when you open a single file and we end up treating all of the system includes as part of your "workspace" which means that we won't skip parsing them.

```
add folder to browse.path: /usr/include/c++/5.4.0
add folder to browse.path: /usr/include/x86_64-linux-gnu/c++/5.4.0
add folder to browse.path: /usr/local/include
add folder to browse.path: /usr/lib/clang/3.8.0/include
add folder to browse.path: /usr/include
add folder to browse.path: /usr/include/x86_64-linux-gnu
add folder to browse.path:      <--  empty path
```

**Comment by MichaelGrupp** [comment]
Glad to help, hope you get this fixed soon. The workaround until then is just to open a folder :wink:

**Comment by bobbrow** [comment]
We released 0.14.0 with a fix for this. Please let us know if you continue to have issues.
